gears/src/x/ante.rs

```diff
@@ -142,6 +147,70 @@ impl<AK: AuthKeeper<SK>, BK: BankKeeper<SK>, SK: StoreKey, GC: SignGasConsumer>
         Ok(())
     }
 
+    fn mempool_fee<M: TxMessage, DB: Database>(
+        &self,
+        ctx: &mut TxContext<'_, DB, SK>,
+        TxWithRaw {
+            tx,
+            raw: _,
+            tx_len: _,
+        }: &TxWithRaw<M>,
+    ) -> anyhow::Result<()> {
+        if !ctx.is_check() {
+            return Ok(());
+        }
+
+        let fee = tx.auth_info.fee.amount.as_ref();
+        let gas = tx.auth_info.fee.gas_limit;
+
+        let min_gas_prices = ctx.options.min_gas_prices();
+
+        if min_gas_prices.is_empty() || min_gas_prices.iter().any(|this| this.amount.is_zero()) {
+            return Ok(());
+        }
+
+        if let Some(fee_coins) = fee {
+            let mut required_fees = Vec::with_capacity(min_gas_prices.len());
+
+            for gp in min_gas_prices {
+                required_fees.push(Coin {
+                    denom: gp.denom,
+                    amount: gp
+                        .amount
+                        .checked_mul(Decimal256::new(gas.into()))?
+                        .to_uint_ceil(),
+                });
+            }
+
+            if !is_any_gte(fee_coins.inner(), &required_fees) {
+                Err(anyhow!(
+                    "insufficient fees; got: {:?} required: {:?}",
+                    fee_coins,
+                    required_fees
+                ))?
+            }
+        } else {
+            Err(anyhow!("rejected. `fee_coins` is None"))?
```

Comment:
This isn't quite right. It's OK for the fee to be `None` provided the required fees are all zero.

Reply:
Isn't this what I implemented? If required fees(min_gas_prices?) is zero or empty we exit from method.

Reply:
You're right I missed the check above where you exit.